File: homework/Truong_Duong/Lesson_9/1011_Capacity_To_Ship_Packages_within_D_Days.js
// Solution for https://leetcode.com/problems/capacity-to-ship-packages-within-d-days/

/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function (weights, days) {
  let totalWeight = 0, maxWeight = weights[0];
  for (const weight of weights) {
    maxWeight = Math.max(maxWeight, weight);
    totalWeight += weight;
  }

  let lo = 0, hi = totalWeight;
  let ans;

  const isDivisible = (weight) => {
    if(weight < maxWeight) return false;

    let portions = 0, i = 0;
    while (i < weights.length) {
      let sumTemp = 0;

      while (i < weights.length && sumTemp + weights[i] <= weight) {
        sumTemp += weights[i];
        i += 1;
      }
      portions += 1;
    }
    return portions <= days;
  }

  while (hi >= lo) {
    const mid = lo + Math.floor((hi - lo) / 2);

    if (isDivisible(mid)) {
      ans = mid;
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }

  return ans;
};

File: homework/thinhbka/lesson-6-binary-search-bst/1011. Capacity To Ship Packages Within D Days.py
def check(capacity,weights,days):
    current = 0
    days-=1
    for e in weights:
        if current + e <= capacity:
            current +=e
        else:
            current = e
            days-=1
    return days >= 0
        
        
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        l = max(weights)
        r = sum(weights)
        while l <= r:
            mid = (l+r)>>1
            if check(mid,weights,days) is True:
                ans = mid
                r = mid-1
            else:
                l = mid+1
        return ans